### PR TITLE
Add partial implementation of CDC subscription FULL mode

### LINKED ISSUE #22500: Add CDC feature
Hi community,

We've planned data synchronization feature before, and we received feature request of CDC (change data capture) from community recently, we'll add CDC feature at next weeks.

CDC feature could be used for ingesting data from Apache ShardingSphere proxy to OLTP, OLAP and MQ system etc.

After several times of discussion with @azexcy and @justbk2015 , review by @tristaZero and @terrymanu , we've got a design for current stage and welcome to feedback.

## Components

- CDC server
- CDC protocol
- CDC client

## CDC server

Supply incremental data. And inventory data is optional.

Part of tasks:

1. Add CDC configuration in`server.yaml`
```
props:
  cdc-server-port: 33071
```

2. Start CDC server and listen on configured port

3. CDC feature is part of pipeline. pipeline-cdc core modules are in L2 layer. New sub-modules will be created.

4. Part of common classes in pipeline-core module will be refactored for CDC

## CDC protocol

Protocol buffers will be used for CDC server and client, for good performance and easy serialization.

## CDC client

Common client classes will be implemented here, e.g.
- Network operation
- Protocol serialization
- API operation
- Basic interface of importing data to target system, it's extensible for easier integration with existing ETL tools

## New modules

- data-pipeline-cdc
  - data-pipeline-cdc-protocol
  - data-pipeline-cdc-core
  - data-pipeline-cdc-client

### PR DESCRIPTION
**Pull request by azexcy**
Fixes #22500.

Changes proposed in this pull request:
  - Add partial implementation of CDC subscription FULL mode 
  - Add unmarshal test

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/prepare/CDCJobPrepare.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        CDCTaskConfiguration taskConfig = jobItemContext.getTaskConfig();
+        // TODO importer and channel requires a new implementation
+        InventoryDumperConfiguration inventoryDumperConfig = new InventoryDumperConfiguration(jobItemContext.getTaskConfig().getDumperConfig());
+        InventoryTaskSplitter inventoryTaskSplitter = new InventoryTaskSplitter(jobItemContext.getSourceDataSource(), inventoryDumperConfig, taskConfig.getImporterConfig());
+        List<InventoryTask> allInventoryTasks = inventoryTaskSplitter.splitInventoryData(jobItemContext);
+        jobItemContext.getInventoryTasks().addAll(allInventoryTasks);
+    }
+    
+    private Collection<InventoryDumperConfiguration> generateInventoryDumperConfigurations(final DumperConfiguration dumperConfig, final PipelineTableMetaDataLoader pipelineTableMetaDataLoader) {
+        Collection<InventoryDumperConfiguration> result = new LinkedList<>();
```
Comment (sandynz) [inline kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/prepare/CDCJobPrepare.java:77]: `generateInventoryDumperConfigurations` method is not used, is it required?

File: kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/prepare/CDCJobPrepare.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * CDC job prepare.
+ */
+@Slf4j
+public final class CDCJobPrepare {
```
Comment (sandynz) [inline kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/prepare/CDCJobPrepare.java:46]: `CDCJobPrepare` could be `CDCJobPreparer`

File: kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/util/JobDataNodeLineConvertUtil.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+/**
+ * Job data node line convert util.
+ */
+public final class JobDataNodeLineConvertUtil {
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/util/JobDataNodeLineConvertUtil.java:35]: Could we add unit tests for JobDataNodeLineConvertUtil?

File: kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/yaml/job/YamlCDCJobConfigurationSwapper.java [hunk 4]
```diff
@@ -17,11 +17,16 @@
...
+import org.apache.shardingsphere.data.pipeline.api.datanode.JobDataNodeLine;
+import org.apache.shardingsphere.data.pipeline.api.datasource.config.impl.ShardingSpherePipelineDataSourceConfiguration;
 import org.apache.shardingsphere.data.pipeline.api.datasource.config.yaml.YamlPipelineDataSourceConfigurationSwapper;
 import org.apache.shardingsphere.data.pipeline.cdc.config.job.CDCJobConfiguration;
 import org.apache.shardingsphere.infra.util.yaml.YamlEngine;
 import org.apache.shardingsphere.infra.util.yaml.swapper.YamlConfigurationSwapper;
 
+import java.util.List;
+import java.util.stream.Collectors;
+
```
Comment (sandynz) [inline kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/yaml/job/YamlCDCJobConfigurationSwapper.java:29]: Could we add unit test for YamlCDCJobConfigurationSwapper?

**Review by sandynz — CHANGES_REQUESTED** [review]

**Comment by sandynz** [comment]
Is it part of the implementation?

File: kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/prepare/CDCJobPrepare.java (hunk 1 above)
Comment (azexcy) [reply kernel/data-pipeline/cdc/core/src/main/java/org/apache/shardingsphere/data/pipeline/cdc/core/prepare/CDCJobPrepare.java:77]: Removed now

**Comment by azexcy** [comment]
> Is it part of the implementation?

Yes, I I changed the title

File: kernel/data-pipeline/core/src/test/java/org/apache/shardingsphere/data/pipeline/core/util/JobDataNodeLineConvertUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        mockDataNodes.put("t_order", dataNodes);
+        mockDataNodes.put("t_order_item", itemDataNodes);
+        List<JobDataNodeLine> jobDataNodeLines = JobDataNodeLineConvertUtil.convertDataNodesToLines(mockDataNodes);
+        assertThat(jobDataNodeLines.size(), is(1));
+        List<JobDataNodeEntry> actualNodeEntry = new ArrayList<>(jobDataNodeLines.get(0).getEntries());
+        assertThat(actualNodeEntry.get(0).getLogicTableName(), is("t_order"));
+        assertThat(actualNodeEntry.get(0).getDataNodes().size(), is(2));
+        assertThat(actualNodeEntry.get(1).getLogicTableName(), is("t_order_item"));
+        assertThat(actualNodeEntry.get(1).getDataNodes().size(), is(1));
+    }
```
Comment (sandynz) [inline kernel/data-pipeline/core/src/test/java/org/apache/shardingsphere/data/pipeline/core/util/JobDataNodeLineConvertUtilTest.java:67]: Looks the content of `assertConvertDataNodesToLinesWith` is the same as `assertConvertDataNodesToLines`

**Review by sandynz — CHANGES_REQUESTED** [review]

**Review by sandynz — APPROVED** [review]

**Pull request merged**